app/web/features/search/useRouteWithSearchFilters.ts

```diff
@@ -0,0 +1,64 @@
+import { useRouter } from "next/router";
+import { useCallback, useEffect, useRef, useState } from "react";
+import SearchFilters, {
+  parsedQueryToSearchFilters,
+  parseSearchFiltersToQuery,
+} from "utils/SearchFilters";
+
+/**
+ * This hook manages a state with search filters and keeps the URL in sync
+ */
+export default function useRouteWithSearchFilters(route: string) {
+  const router = useRouter();
+  const [active, setActive] = useState<SearchFilters>(() =>
+    parsedQueryToSearchFilters(router.query)
+  );
+  const pending = useRef<SearchFilters>(active);
+
+  const expectedRouteWithQuery = `${route}?${parseSearchFiltersToQuery(
+    active
+  )}`;
+  const routerQueryIsReady = router.isReady;
```

Comment:
Didn't know about this when switching from cra to next! Did you encounter any bugs?

Reply:
yeah, when navigating with React Router to `earch?location=...`, `useEffect` below was being invoked when the router wasn't ready, and so `router.query` was empty. This didn't happen when opening the URL directly in a new tab or something (for example, opening this link: https://couchers.org/search?location=Barcelona%2C+Catalonia%2C+Spain&lat=41.3828939&lng=2.1774322)

It's tricky. React Router added this flag not so long ago I think. I came across issues from not checking it in another project recently, I think it's related to Next.js SSR, but I haven't dug deeper yet